lib/common/common/src/cpu.rs

```diff
@@ -1,7 +1,196 @@
+        Self {
+            semaphore: Arc::new(Semaphore::new(cpu_budget)),
+            cpu_budget,
+        }
+    }
+
+    /// For the given desired number of CPUs, return the minimum number of required CPUs.
+    fn min_permits(&self, desired_cpus: usize) -> usize {
+        desired_cpus.min(self.cpu_budget).div_ceil(2)
+    }
+
+    /// Try to acquire CPU permit for optimization task from global CPU budget.
+    ///
+    /// The given `desired_cpus` is not exact, but rather a hint on what we'd like to acquire.
+    /// - it will prefer to acquire the maximum number of CPUs
+    /// - it will never be higher than the total CPU budget
+    /// - it will never be lower than `min_permits(desired_cpus)`
+    pub fn try_acquire(&self, desired_cpus: usize) -> Option<CpuPermit> {
+        // Determine what number of CPUs to acquire based on available budget
+        let min_required = self.min_permits(desired_cpus) as u32;
+        let num_cpus = self.semaphore.available_permits().min(desired_cpus) as u32;
+        if desired_cpus == 0 || num_cpus < min_required {
+            return None;
+        }
+
+        // Try to acquire selected number of CPUs
+        let result = Semaphore::try_acquire_many_owned(self.semaphore.clone(), num_cpus);
+        let permit = match result {
+            Ok(permit) => permit,
+            Err(TryAcquireError::NoPermits) => return None,
+            Err(TryAcquireError::Closed) => unreachable!("Cannot acquire CPU permit because CPU budget semaphore is closed, this should never happen"),
+        };
+
+        Some(CpuPermit::new(num_cpus, permit))
+    }
+
+    /// Check if there is enough CPU budget available for the given `desired_cpus`.
+    ///
+    /// This checks for the minimum number of required permits based on the given desired CPUs,
+    /// based on `min_permits`. To check for an exact number, use `has_budget_exact` instead.
+    ///
+    /// A desired CPU count of `0` will always return `true`.
+    pub fn has_budget(&self, desired_cpus: usize) -> bool {
+        self.has_budget_exact(self.min_permits(desired_cpus))
+    }
+
+    /// Check if there are at least `budget` available CPUs in this budget.
+    ///
+    /// A budget of `0` will always return `true`.
+    fn has_budget_exact(&self, budget: usize) -> bool {
+        self.semaphore.available_permits() >= budget
+    }
+
+    /// Wait until we have CPU budget available for the given number of desired CPUs.
+    ///
+    /// Waits for at least the minimum number of permits based on the given desired CPUs. For
+    /// example, if `desired_cpus` is 8, this will wait for at least 4 to be available. See
+    /// [`Self::min_permits`].
+    ///
+    /// - `1` to wait for any CPU budget to be available.
+    /// - `0` will always return immediately.
+    ///
+    /// Uses an exponential backoff strategy up to 10 seconds to avoid busy waiting.
+    pub async fn wait_until_budget(&self, desired_cpus: usize) {
```

Comment:
This function waits until the budge is available but does not guarantee that the caller will be able to use it?
I am wondering what is happening when we have several tasks waiting on the same budge.

Reply:
yeah, I guess I would rather name it "notify on budget change" in this case.

Reply:
> This function waits until the budge is available but does not guarantee that the caller will be able to use it?
I am wondering what is happening when we have several tasks waiting on the same budge.

You're right. We cannot do this in an atomic way.

The idea is that we don't really care whether we can eventually claim it, because we'd just try it again later. It is more like a notification: hey, you can give it another try.

I like the name suggestion, I'll make that change now.